File: src/main/kotlin/services/validation/Validation.kt
package com.kitledger.services.validation

import io.konform.validation.ValidationResult as KonformValidationResult
import io.konform.validation.Valid
import io.konform.validation.Invalid

enum class ValidationErrorType {
    DATA,
    STRUCTURE
}

/**
 * Represents a validation error.
 */
data class ValidationError(
    val type: ValidationErrorType,
    val path: String?,
    val message: String
)

/**
 * Represents the result of a validation.
 */
sealed class ValidationResult<out T> {
    data class Success<out T>(val data: T) : ValidationResult<T>()
    data class Failure(val errors: List<ValidationError>) : ValidationResult<Nothing>()
}

fun <T> KonformValidationResult<T>.toNativeValidationResult(): ValidationResult<T> {
    return when (this) {
        is Valid -> ValidationResult.Success(this.value)

        is Invalid -> {
            val nativeErrors = this.errors.map { konformError ->
                ValidationError(
                    type = ValidationErrorType.DATA,
                    path = konformError.dataPath.ifEmpty { null }, // Use null for root-level errors
                    message = konformError.message
                )
            }
            ValidationResult.Failure(nativeErrors)
        }
    }
}

sealed interface ActionResult<out T> {
    /** The action was successful and returned data. */
    data class Success<out T>(val data: T) : ActionResult<T>

    /** The action was not attempted because the input data was invalid. */
    data class ValidationFailure(val errors: List<ValidationError>) : ActionResult<Nothing>

    /** The action was attempted but failed during execution (e.g., database error). */
    data class ActionFailure(val message: String) : ActionResult<Nothing>
}

File: packages/core/src/validation.ts
import { type BaseIssue, type IssuePathItem } from "valibot";

export type ValidationError = {
	type: "structure" | "data";
	path: string | null;
	message: string;
};

export function parseValibotIssues<T>(issues: BaseIssue<T>[]): ValidationError[] {
	return issues.map((issue) => ({
		type: "structure",
		path: issue.path ? issue.path.map((p: IssuePathItem) => p.key).join(".") : "",
		message: issue.message,
	}));
}

export type ValidationSuccess<T> = {
	success: true;
	data: T;
};

export type ValidationResult<T> = {
	success: boolean;
	data?: T;
	errors?: ValidationError[];
};

export type ValidationFailure<T> = {
	success: false;
	data?: T;
	errors?: ValidationError[];
};

export function isValidationFailure<T extends object, U>(
	result: T | ValidationResult<U>,
): result is ValidationResult<U> {
	return "errors" in result;
}
